Add timeout to saltnado's event listener

jacksontj opened the pull request:
Fixes #21707

The issue is basically that the master has seen N minions-- and not all N are currently connected. The publish job returns to the salt-api that N minions got the job-- and it dutifilly waits for all the returns. With this we use the same ping timeout that the CLI does while waiting for returns. So once all minions are no longer running the job we will return.

jacksontj committed:
Add timeout to saltnado's event listener

Fixes #21707

The issue is basically that the master has seen N minions-- and not all N are currently connected. The publish job returns to the salt-api that N minions got the job-- and it dutifilly waits for all the returns. With this we use the same ping timeout that the CLI does while waiting for returns. So once all minions are no longer running the job we will return.

jacksontj commented:
@rallytime or @jfindlay should I open this PR against 2015.2 as well? Or will this get merged forward auto-magically?

jfindlay commented:
Merge forwards happen whenever Colton gets a few free minutes; :-) usually once every 2-4 days.

jacksontj commented:
@jfindlay To make sure it isn't missed in 2015.2 I just opened #21756 :) I figure that'll be easier

jfindlay commented:
Sure, thanks @jacksontj, great work as usual.

jacksontj committed:
Massive speedup to saltnado

This mixin was creating 3 localclients and 1 runner client per request-- which takes ~6s. Now we build the set of clients once per process

Conflicts:

	salt/runner.py

whiteinge commented:
@jacksontj ooc, could you lean on the existing timeout semantics in LocalClient() instead of implementing it here? We moved the `LocalClient().cmd()` method to use the same timeout behavior as the CLI for the 2014.7 release.

jacksontj commented:
While I'm in here we'll go ahead and backport that big performance fix ;)

jacksontj commented:
@whiteinge Depends on what you mean by that-- this is basically making it have the same timeout semantics as localclient and the CLI (before it was a mess). The only real change here was that we assumed that the Publisher's list was exhaustive and accurate-- so we would never timeout waiting for a ping return of a minion-- which in the case of a dead one means you'll wait until someone turns it back on.

If you mean "can we use the same code" that gets more complicated-- since saltnado is all ioloop/async stuff it doesn't quite work the same-- but the logic here is equivalent-- just lots less code ;) I have a branch on my github where I'm moving most of the internals over to tornado stuff-- so we'll be able to consolidate the timeout code once that's done.

whiteinge commented:
Roger that (sorry for the brevity in my last comment; wild day today). You're very right that it was a mess -- `cmd()` and Salt's CLI had badly diverging behavior. In 2014.7 I moved `LocalClient().cmd()` to use `get_cli_event_returns` so it should be identical to the CLI now and hopefully not a mess. (Or at least a centrally-managed mess.)

Good point that Tornado is coming at this from an async perspective. I don't know how well the existing `get_iter_returns` (which `get_cli_event_returns` wraps) works with that kind of system (if at all). I know some work was done to get it Gevent-friendly a while back but that's a very different async strategy.

Nevermind me. :walking:

jacksontj commented:
Lol, no worries ;) it is a bit of a mess. Now (as you mentioned) its a central mess, but a mess nonetheless. This implementation has the same outcome but just has less code-- plan is to replace the ugly mess with this code (or the equivalent), but it's part of a larger re-write ;) We'll get there :)


It might be nice to make a test case for testing command returns-- that way we can test that cmd(), api, and cli have the same outcome-- since we can't really gaurantee that they'll use the same code forever-- specifically it is probably worth making a minion in the integration suite which isn't on (since thats the case we are testing).

whiteinge commented:
+1

cachedout commented on salt/runner.py:
```diff
@@ -53,6 +53,7 @@ def cmd(self, fun, arg, pub_data=None, kwarg=None):
 
         .. code-block:: python
 
+<<<<<<< HEAD
```
Could you pull this out please @jacksontj ?

jacksontj committed:
Cleanup merge

jacksontj commented:
@cachedout Done

cachedout commented:
Thank you @jacksontj

baweaver commented:
Could we look at potentially adding something to the README section on this one? I assume clients would just send something to the equivalent of ``-d timeout=n``?

jacksontj commented:
@whiteinge Now you have me all worried about the timeouts not matching up. I'll take a third look at this later today ;)

isbm commented on salt/netapi/rest_tornado/saltnado.py:
```diff
@@ -294,15 +290,30 @@ def get_event(self,
         self.request_map[request].append((tag, future))
 
+        if timeout:
+            timeout_future = tornado.ioloop.IOLoop.current().call_later(timeout, self._timeout_future, tag, future)
```
Sorry for the nitpick, but this probably wants to be wrapped a little. :)

isbm commented on salt/netapi/rest_tornado/saltnado.py:
```diff
@@ -896,7 +910,7 @@ def _disbatch_local(self, chunk):
                 if ping_id not in chunk_ret and ping_id not in minions_remaining:
                     minions_remaining.append(ping_id)
-                ping_event = self.application.event_listener.get_event(self, tag=ping_tag)
+                ping_event = self.application.event_listener.get_event(self, tag=ping_tag, timeout=self.application.opts['gather_job_timeout'])
```
This wants to be wrapped too.

isbm commented on salt/netapi/rest_tornado/saltnado.py:
```diff
@@ -881,7 +895,7 @@ def _disbatch_local(self, chunk):
 
         ret_event = self.application.event_listener.get_event(self, tag=ret_tag)
-        ping_event = self.application.event_listener.get_event(self, tag=ping_tag)
+        ping_event = self.application.event_listener.get_event(self, tag=ping_tag, timeout=self.application.opts['gather_job_timeout'])
```
And this line is too long.

jacksontj commented:
@whiteinge Yup, we didn't handle the case where a minion started with the job, then lost it (minion died, job stopped, w/e). #21811 includes the fixes for that-- which I think ends up being a nicer interface-- since its a single future that you are returned :)
